Description: CVE-2026-23865 - Check for overflow in array size computation
 An integer overflow in ft_var_load_item_variation_store when computing
 the delta set array size (regionIdxCount * itemCount) may allow for an
 out-of-bounds read when parsing HVAR/VVAR/MVAR tables in OpenType
 variable fonts. This patch adds an overflow check on the multiplication
 before passing the result to FT_NEW_ARRAY.
Origin: upstream, https://gitlab.com/freetype/freetype/-/commit/fc85a255849229c024c8e65f536fe1875d84841c
Author: Werner Lemberg <wl@gnu.org>
Last-Update: 2026-03-05

--- a/src/truetype/ttgxvar.c
+++ b/src/truetype/ttgxvar.c
@@ -449,6 +449,7 @@
     FT_Error   error;
     FT_UShort  format;
     FT_ULong   region_offset;
+    FT_UInt    delta_set_size;
     FT_UInt    i, j, k;
     FT_UInt    wordDeltaCount;
     FT_Bool    long_words;
@@ -618,7 +619,19 @@
       /*                                                                 */
       /* On output, deltas are expanded to `regionIdxCount` shorts each. */
-      if ( FT_NEW_ARRAY( varData->deltaSet,
-                         varData->regionIdxCount * varData->itemCount ) )
+      /* Check for overflow (we actually test whether the     */
+      /* multiplication of two unsigned values wraps around). */
+      delta_set_size = varData->regionIdxCount * varData->itemCount;
+      if ( varData->regionIdxCount                                &&
+           delta_set_size / varData->regionIdxCount !=
+             varData->itemCount                                   )
+      {
+        FT_TRACE2(( "ft_var_load_item_variation_store:"
+                    " bad delta set array size\n" ));
+        error = FT_THROW( Array_Too_Large );
+        goto Exit;
+      }
+
+      if ( FT_NEW_ARRAY( varData->deltaSet, delta_set_size ) )
         goto Exit;

       /* the delta set is stored as a 2-dimensional array of shorts */